Add `postgrest-loadtest` based on vegeta

wolfgangwalther opened the pull request:
This is my take on #1609. As discussed in that PR, I ran a few tests with [vegeta](https://github.com/tsenart/vegeta).

Compared to locust, as initially tested in #1609, I identified the following advantages:
- Support for unix sockets out of the box
- Running single-threaded is possible (needed for CI with limited no of cores)
- Supports HTTP/2, once we need it
- No boilerplate at all for tests - using the plain-text HTTP message format
- Written in a compiled language for more req/s

I have only written one very basic test, so far. We are not doing too much with the results either, yet. However, I have added a nice framework around the test runner. We have:
- `postgrest-loadtest` to run the loadtests in the current working tree
- `postgrest-loadtest-against <commit-ish>` to run the loadtests twice for A/B testing. Once on the current working tree and once on the branch that is given by the first argument. It defaults to main, that means `postgrest-loadtest-against` will give you a test run comparing your current branch against main. This, with some fancy number crunching of results, could be a great tool for CI.
- As a side-effect of implementing that, I learned about `git worktree` (quite awesome!) and implemented `postgrest-with-git <commit-ish> <sub-command>`. This will checkout commit-ish in a temporary directory and run the sub-command in it. This not only allows to compare loadtests on different branches, but also things like:
  - I am currently working on a fancy new nix feature (this one here) and [some reviewer suggests a change to one of my open PRs](https://github.com/PostgREST/postgrest/pull/1808#issuecomment-817993627). To make the change without doing anything to my current working tree, I just ran `postgrest-with-git remove-setup-hs nix-shell`. This started a new nix-shell on that branch. Made the change, pushed, `exit`, done. Ok, that's really what you would do with `git worktree` anyway - but I didn't know, so that was quite an awesome thing to do :D.
  - Remo is going on another refactor-spree? Not sure what that does to performance/tests/whatever? Ah, just a quick `postgrest-with-git remo/untangle-the-world postgrest-test-spec-all` will answer that.
  - Just developed some cool new nix tooling, like we always do, but it's not yet committed and you want to run this tool against a different branch? `postgrest-with-git that-other-branch postgrest-my-new-tool` has you covered, because even though it runs in the other branch, it still takes the `postgrest-` command from the current instance of nix-shell. Of course, you can do by just running `nix-shell` on one branch and then checking out the other branch. Once you have uncommitted changes, it becomes a bit complicated with possible merge conflicts, though... `postgrest-with-git` has you covered!

After writing this down, I am not sure anymore what the main feature of this PR was... ;)

All those tools are available behind `nix-shell --arg loadtest true` right now - although the only dependency is now vegeta. So we should just enable it by default, I guess?

There are a few smaller todos in code, but the bigger ones are (mostly taken from #1609):
- [x] Add to CI
- [x] Create human readable report
- [ ] Validate consistency across runs in CI
- [ ] Crunch the numbers, compare to a threshold and report back to the PR via GitHub Status API
- [x] Add lots of tests
- [x] Add proper bash completion

Any feedback more than welcome!

monacoremo commented:
Awesome!

> So we should just enable it by default, I guess?

Yes, absolutely!

steve-chavez commented:
> postgrest-loadtest-against <commit-ish>
> postgrest-with-git remo/untangle-the-world postgrest-test-spec-all

I'd love to have those commands available :D

Amazing work Wolfgang :100: :1st_place_medal:  

> No boilerplate at all for tests - using the plain-text HTTP message format

I like how simple the HTTP test is. One thing though, would it be possible to add a random id(as a query param) that changes on every request? Similar to what I do [here](https://github.com/supabase/benchmarks/blob/master/postgrest/k6/GETSingle.js#L58-L59).

wolfgangwalther commented:
> > No boilerplate at all for tests - using the plain-text HTTP message format
> 
> I like how simple the HTTP test is. One thing though, would it be possible to add a random id(as a query param) that changes on every request? Similar to what I do [here](https://github.com/supabase/benchmarks/blob/master/postgrest/k6/GETSingle.js#L58-L59). This would be useful for future load tests we might add.

I guess this would be possible, but would require a bit of overhead around it. Not as simple anymore, for sure. What advantage would you expect when we did randomize the id?

wolfgangwalther commented:
I added a few tests regarding headers, because I wanted to know more about https://github.com/PostgREST/postgrest/pull/1794#issuecomment-817572919.

steve-chavez commented:
> What advantage would you expect when we did randomize the id?

Fairer tests I thought. Might not be that relevant on GET, because during my load tests on cloud, I noticed the same req/s with both same id vs random ids. Testing PATCH on different ids might make a difference in results(though maybe it's more of a postgresql concern, see [here](https://github.com/supabase/benchmarks/issues/2#issuecomment-738870014)).

wolfgangwalther commented:
> > What advantage would you expect when we did randomize the id?
> 
> Fairer tests I thought. Might not be that relevant on GET, because during my load tests on cloud, I noticed the same req/s with both same id vs random ids. Testing PATCH on different ids might make a difference in results(though maybe it's more of a postgresql concern, see [here](https://github.com/supabase/benchmarks/issues/2#issuecomment-738870014)).

Hm. I was about to argue that compared to the "whole-system-k6-tests" you're running, where PostgreSQL and it's performance is very important, we don't need to care about PostgreSQL in `postgrest-loadtest`. To a certain degree this is true, because when we're testing the efficiency of our haskell implementation, it wouldn't matter. But when we're looking at things like https://github.com/PostgREST/postgrest/pull/1794#issuecomment-818847223, we obviously care about what PostgreSQL does.

So we should probably draw the line somewhere around the fact, where the performance problem can be fixed:
- If we decided to do something about the GUC interface in general and improve performance, we should add some tests, e.g. with many headers, to prevent us from having a regression here.
- Reading through the `PATCH` issue you've linked, I understand that the fix really is in to be done in the PostgreSQL configuration - not in our code. So for that specific case, I would not implement random ids, because we wouldn't test any possible regression here.

Adding randomness also potentially makes test runs less comparable, if we don't counter that with longer runtimes to balance out the random part. We need to keep in mind that the more tests we run with the loadtest, the longer we need to run the test to get proper results per test-case. I don't think we need to be super careful, but we shouldn't add every possible test-case either. The goal should be to have tests in place for all major paths through PostgREST - and then some additional tests for specific issues we've encountered and fixed to avoid regressions.

WDYT?

steve-chavez commented:
> Adding randomness also potentially makes test runs less comparable, if we don't counter that with longer runtimes to balance out the random part. We need to keep in mind that the more tests we run with the loadtest, the longer we need to run the test to get proper results per test-case.

Very true. Let's keep random ids out until they're really needed.

> The goal should be to have tests in place for all major paths through PostgREST - and then some additional tests for specific issues we've encountered and fixed to avoid regressions.

Yes, Fully agree.

wolfgangwalther commented:
Last night I felt brave and tried running `postgrest-loadtest-against v7.0.1` to see where we're at. Had to made a few changes to support older versions, because earlier commits don't have environment variable support. That part is not pushed, yet, because I took a few shortcuts with hardcoded paths etc...

Anyway the result was... surprising. v7.0.1 seemed to be muuch faster than current HEAD. I wanted to see what was going on and wrote a small script to run `postgrest-loadtest` against every commit since v7.0.1 in order. To get an idea of how consistent the measurements are going to be, I made each run 3 times, with 1s, 5s and 25s duration respectively. I only ran one very simple `GET` request, nothing more. The numbers are given in hypothetical req/s (based on minimum latency). The higher, the better.

<details>
  <summary>Here's the data</summary>

hash|date|1s|5s|25s|subject
-- | -- | -- | -- | -- | --
222a5301|2020-05-18T16:50:53-05:00|3543|3552|3671|Fix Dockerfile
67344c8e|2020-05-19T13:59:33-05:00|3565|3572|3622|cabal: fix upload to hackage (#1532)
289bb66f|2020-05-20T11:49:11-05:00|3543|3639|3610|Reduce size of the static docker image by patching the Nix openssl package (#1528)
e6874c86|2020-05-20T14:10:53-05:00|3637|3631|3668|Fix UNKNOWN being shown as git commit hash (#1533)
48c9ac36|2020-05-20T21:30:20-05:00|3553|3635|3675|Bump nixpkgs and simplify/clean up nix setup (#1529)
3a1844ec|2020-05-21T14:47:07-05:00|3683|3651|3727|Fix the static build (#1534)
d4aba5cb|2020-05-22T14:30:23-05:00|3719|3578|3669|Add notice about single Linux static executable
08186ea5|2020-05-25T11:27:35-05:00|3724|3687|3682|Nixify CircleCI setup (#1535)
784ebe57|2020-05-26T10:43:14-05:00|3517|3579|3666|Allow aeson 1.5 (#1537)
69b09e31|2020-05-28T12:45:14-05:00|3552|3552|3656|Nixify io and memory tests (#1538)
0f0d6179|2020-06-05T13:01:36-05:00|3437|3534|3582|Allow http status override through response.status guc (#1541)
8588a42a|2020-06-05T13:40:03-05:00|3517|3556|3595|Fix broken links in .github dir
24064f86|2020-06-05T13:57:47-05:00|3549|3510|3529|github: CHANGELOG reminder in PR template
43d71e95|2020-06-24T19:00:02-05:00|3515|3546|3588|Allow schema cache reloading with NOTIFY (#1542)
a5bc2933|2020-07-03T13:54:39-05:00|3462|3582|3621|add test cases for singular minimal
55b4f4fb|2020-07-03T17:23:10-05:00|3438|3543|3602|Fix expired JWTs starting an empty transaction
343e41c5|2020-07-08T14:36:40-05:00|3525|3548|3570|Location header improvements(#1475)
896b79f0|2020-07-13T11:30:16-05:00|3557|3506|3539|refactor: separate reading file from parsing it
96a16a37|2020-07-13T11:30:16-05:00|3538|3522|3545|refactor: move loadDbUriFile/SecretFile to Config
0ff05edd|2020-07-13T11:30:16-05:00|3492|3542|3523|refactor: move parseSecret out of App.postgrest
e272ea47|2020-07-13T11:30:16-05:00|3478|3612|3632|refactor: move logStdout/corsPolicy to Middleware
e8b4e377|2020-07-13T11:30:16-05:00|3592|3536|3524|Allow config file reloading with SIGUSR2
1f6a824d|2020-07-13T18:47:45-05:00|3597|3525|3572|Bump nixpkgs pin to unstable as of 2020-07-13
6b2767d3|2020-07-16T12:09:40-05:00|3414|3535|3617|"Additionally allow ""bearer"" instead of only ""Bearer"" in Authorization… (#1558)"
18984792|2020-07-16T13:39:25-05:00|3574|3578|3589|Schema cache reload with zero downtime (#1559)
8e4687fb|2020-07-16T16:05:57-05:00|3477|3572|3616|Return 405 Method not Allowed for GET of volatile RPC instead of 500 (#1560)
a3f4548a|2020-08-14T21:12:58-05:00|3497|3588|3570|Allow base64-bytestring 1.2 and optparse-applicative 0.16
bd2160db|2020-08-14T21:12:58-05:00|3468|3581|3552|Make cabals bound more consistent
7af54c58|2020-09-09T19:08:16-05:00|3390|3590|3512|Upgrade nixpkgs to current master and add an upgrade checklist (#1579)
807dae17|2020-09-22T11:45:58-05:00|3501|3549|3582|Allow wai-extra 3.1
17f04a88|2020-09-27T16:04:47-05:00|3498|3560|3593|Allow hspec-wai/hspec-wai-json 0.11 (#1595)
d395bb60|2020-10-03T10:11:41-05:00|3551|3546|3537|Fix error messages on connection failure for postgres on localized Windows
e4e84e57|2020-10-05T13:47:53-05:00|3457|3612|3541|Upgrade nixpkgs and add Postgres 13 support
741f017a|2020-10-05T13:47:53-05:00|3561|3555|3558|Use Postgres 13 in tests/docker-compose.yml
60398ad5|2020-10-06T12:45:40-05:00|3469|3497|3558|Allow wai-middleware-static 0.9
e9efcc70|2020-10-06T14:21:46-05:00|3577|3604|3601|Add log-level config
f6b6abe7|2020-10-06T14:21:46-05:00|3465|3606|3629|Add error log level
d9a608d9|2020-10-06T14:21:46-05:00|3632|3611|3628|Add warn log level
ccad9eb9|2020-10-06T14:21:46-05:00|3578|3597|3643|Change default log-level, from info to error
99ecc7af|2020-10-13T10:50:19-05:00|3493|3546|3597|Update stack.yaml.lock (#1616)
2eb80838|2020-10-14T09:10:47-05:00|3578|3625|3673|Simplify Nix CI jobs
5bb670b7|2020-10-14T09:26:35-05:00|3462|3646|3582|fix appveyour build script
f5331d1a|2020-10-14T09:26:35-05:00|3546|3641|3617|fix build error on windows
b21a0f61|2020-10-14T10:49:47-05:00|3515|3572|3604|fix travis build timeout swallowing error code
29858b8d|2020-10-14T22:25:42-05:00|3501|3620|3645|improve cirrus build time by properly using cabal cache
719c4abb|2020-10-15T08:43:28-05:00|3558|3580|3629|Drop support for postgres 9.4
5f33f010|2020-10-16T10:24:41-05:00|3481|3558|3555|correct issue numbers in changelog
78097088|2020-10-16T11:15:20-05:00|3545|3616|3623|ci: run spec tests against all pg versions, even if one of them fails (#1628)
b7b66b60|2020-10-16T13:53:43-05:00|3496|3546|3557|ci: remove io tests from stack-test because of random segmentation faults (#1627)
7e3e19ac|2020-10-17T15:01:06-05:00|3576|3705|3558|refactor: untangle addFiltersOrdersRanges
d1d0c677|2020-10-17T15:01:06-05:00|4095|4084|4131|perf: change Text queries to ByteString
eed0d3a6|2020-10-22T12:05:25-05:00|3964|4067|4058|ci: add coveralls.io integration
06e85357|2020-10-26T17:49:51-05:00|3992|4030|4076|add tests to pass regular arrays properly to RPCs
18cc214c|2020-10-26T17:49:51-05:00|4035|4011|4145|refactor moving findProc into ApiRequest to allow parsing parameters differently by proc
302d4e15|2020-10-26T17:49:51-05:00|3975|4036|4105|Allow calling variadic functions with repeated query params or JSON array in body
04eaeec7|2020-10-29T18:53:24-05:00|3921|4051|4144|refactor: reorder PgVersion(95/96)Spec tests
a6696f3b|2020-10-29T18:53:24-05:00|3951|4042|4046|refactor: add Hasql arrayColumn decoder
2798ced9|2020-10-29T18:53:24-05:00|4048|3933|4116|Fix charset=utf-8 appending to binary output
8618ffa5|2020-10-29T18:53:24-05:00|3969|4160|4058|perf: shortcut for proc with no variadic arg
4b4a622a|2020-10-30T10:06:53-05:00|3940|4085|4078|fix non-variadic repeated param in variadic function
efc725fb|2020-11-02T13:22:01-05:00|4014|4073|3991|run stylish-haskell v0.12.1.0
a7403fec|2020-11-03T10:43:06-05:00|4018|4031|4120|refactor: split InsertSpec into InsertSpec and UpdateSpec
122fea15|2020-11-03T10:43:06-05:00|3763|4020|4089|Add test to insert into a view and expect location header return
9f8d1af2|2020-11-10T22:47:58-05:00|3904|4013|4055|nix: add postgrest-docker-login wrapper func
63849f12|2020-11-10T22:47:58-05:00|3838|4008|4146|circleci: make release job a machine
139f9b98|2020-11-10T22:47:58-05:00|3833|3991|4001|nix: fix postgrest-release-dockerhubdescription
5f3b5815|2020-11-10T22:47:58-05:00|4031|4020|4046|appveyor: add nightly release
8fd9a9ab|2020-11-10T22:47:58-05:00|3929|4021|4043|travisci: add nightly release
51b43f56|2020-11-10T22:47:58-05:00|4015|4038|4069|nix: change release scripts to use a version arg
5d140f6f|2020-11-10T22:47:58-05:00|3938|4060|4047|circleci: add nightly release
814da058|2020-11-10T22:47:58-05:00|3988|4033|4036|cirrusci: add nightly release
b6d8d89b|2020-11-10T22:47:58-05:00|4020|4086|4086|ci: add git sha with hours/minutes to nightly
38308875|2020-11-10T22:47:58-05:00|3784|4030|4073|ci: add nightly version to cabal file
ed8bf827|2020-11-13T14:01:58-05:00|3997|3961|4033|Update BACKERS.md
65968b53|2020-11-13T14:12:28-05:00|3922|4027|4097|circleci: fix nightly release
6e04fe74|2020-11-15T14:58:57-05:00|3877|4074|4065|Enable embedding through multiple layers of views recursively (#1625)
3f690ec7|2020-11-17T18:49:02-05:00|3847|4046|4061|Removing single column limit from join table M2M mapping detection. (#1593)
ca7ffd0a|2020-11-20T09:24:32-05:00|3979|4064|4045|Fix RPC return type handling for domains with composite base type
b0915863|2020-11-20T09:24:32-05:00|3892|4005|4064|Refactor tests: renaming json table to prevent hiding of native json type
e08bb3a1|2020-11-20T09:24:32-05:00|3932|4036|4047|Fix overloading of functions with unnamed arguments (specifically for prefer params=single-object)
2c52b96e|2020-11-20T09:24:32-05:00|3993|4091|4047|Refactor tests: replace str with json QuasiQuoter where appropriate
d91f47e5|2020-11-20T10:48:47-05:00|3901|4035|4022|Remove executable bit on .nix and .sql file
944efb3a|2020-11-20T11:51:49-05:00|3958|4018|4029|Split StructureSpec into OpenApiSpec and OptionsSpec
0d5520d9|2020-11-21T15:50:27-05:00|3893|4034|4054|Update README.md
698fac8f|2020-11-22T14:48:48-05:00|3859|4056|4043|Add test to ignore functions unnamed arguments for rpc disambiguation, resolves #1638
dbf99c6a|2020-11-22T18:21:06-05:00|3930|4035|4039|Add support for Prefer tx=rollback
34253520|2020-11-23T17:40:04+01:00|4055|4068|4063|Clean up postgrest --help output
b13e95ae|2020-11-23T22:43:49+01:00|3841|4094|4048|Refactor tx-... config options to single tx-end option
8e58b56d|2020-11-23T19:00:50-05:00|3911|4034|4098|refactor: Use hasql-dynamic on RPC/POST/PUT/PATCH
4bd5e6bd|2020-11-23T19:00:50-05:00|5530|5812|5846|perf: enable prepared statements for GET
787973f3|2020-11-23T19:00:50-05:00|5573|5683|5777|Add db-prepared-statements config
9adec12a|2020-11-23T19:00:50-05:00|5603|5678|5766|cirrusci: add build timeout
07cb47e6|2020-11-30T22:21:19+01:00|5542|5608|5726|Improve dev tools in the Nix environment (#1666)
f2cb9174|2020-12-01T00:37:45+01:00|5442|5693|5692|Use cabal v2-exec to run io tests, include io tests in nix-shell by default; remove obsolete ncat dependency (#1673)
1a5f6cce|2020-12-01T00:42:21+01:00|5549|5577|5811|Add docker files to run nix dev environment on Windows
de5742fc|2020-12-01T00:42:21+01:00|5513|5895|5781|Make nix the one and only dev environment; remove obsolete tooling
10a70d4e|2020-12-01T22:13:41+01:00|5446|5690|5812|add autocompletion for postgrest-watch
609c9aea|2020-12-03T18:54:33+01:00|5623|5683|5772|Make transaction-rollback=true the default for test-suite (#1663)
7d6d0158|2020-12-03T22:32:07+01:00|5552|5924|5766|improve error output for io-tests and with_tmp_db
d218a9ea|2020-12-05T16:47:50+01:00|5717|5602|5690|dev: improve postgrest-watch experience
eb46cf26|2020-12-06T13:33:25+01:00|5549|5696|5769|added: cli option --dump-config prints loaded config and exits
8979442e|2020-12-06T13:33:25+01:00|5548|5649|5742|dev: add colors to io tests
ab337599|2020-12-06T20:49:48+01:00|5296|5659|5756|ci: fix cirrus low memory error
ed58511d|2020-12-06T21:59:03+01:00|5407|5579|5794|feat: renamed config options with prefixes; added aliases for old names
9254f119|2020-12-06T21:59:32+01:00|5543|5762|5837|fix: implement robust parsing of boolean config values
7069bb3c|2020-12-07T20:09:18-05:00|4533|4776|4777|refactor: Change SET LOCAL gucs to set_config
11d62a80|2020-12-07T20:09:18-05:00|5431|5538|5641|Prepared statement for set_config
ebd474a7|2020-12-09T14:20:00-05:00|5380|5580|5610|fix: retry connection on failed schema cache load (#1685)
093fd3c8|2020-12-10T19:48:05+01:00|5569|5493|5554|fix: fix embedding through views with subqueries inside function calls
0c25f128|2020-12-10T19:55:42+01:00|5341|5542|5608|fix: fix output of RPCs returning scalar values with multiple rows
71fa8793|2020-12-13T16:19:26+01:00|5355|5639|5693|Migrate io-tests from bash to pytest
a52c114f|2020-12-13T22:13:05+01:00|5499|5412|5576|add test for spec idempotence
dbe3c163|2020-12-14T00:42:03+01:00|5369|5543|5584|run spec tests in parallel
fe096377|2020-12-17T17:05:03-05:00|5256|5502|5586|catch ReadTimeout in IO tests
bf141ca1|2020-12-20T10:22:52+01:00|5391|5661|5595|feat: Added --dump-schema CLI option to dump JSON of dbStructure schema cache.
9cfc66a6|2020-12-23T09:39:36-05:00|5368|5653|5524|upgrade nixpkgs and apply related fixes
f2f639e4|2020-12-23T09:39:36-05:00|5403|5517|5591|update stack to lts-16.26
b7fc393e|2020-12-23T19:39:40+01:00|5453|5573|5723|feat: Read config directly from environment variables
add326a7|2020-12-23T19:39:40+01:00|5352|5497|5816|ci: skip postgrest-release-dockerhubdescription for nightly builds
69b459e7|2020-12-23T19:39:40+01:00|5414|5489|5563|Use environment variables for configuration in developer tooling
2cbe1ba9|2020-12-23T19:39:40+01:00|5400|5555|5580|feat: Add --example cli option to show example config file
56557c9c|2020-12-23T19:39:40+01:00|5534|5560|5731|ci: remove broken travis coverage job
b5185de7|2020-12-24T14:44:01+01:00|5364|5557|5557|"nix: add ""cd $rootdir"" to all shell scripts via checked-shell-script"
2425cdda|2020-12-24T16:09:25+01:00|5530|5526|5675|improve io-tests
96478ed0|2020-12-28T13:23:12+01:00|5325|5530|5664|Add docs to all checked shell scripts in nix
8e025519|2020-12-29T16:07:07+01:00|5446|5484|5548|"Fix ""unused-imports"" warning for windows build"
20156883|2020-12-29T23:58:37+01:00|5353|5525|5527|Fix spec test concurrency issue related to tx=commit on simple_pk
c7f0d423|2020-12-30T13:22:25+01:00|3642|3744|3725|Add postgrest-coverage to show and upload hpc reports to codecov
568477b6|2020-12-30T14:51:58+01:00|3683|3734|3805|Fix io tests not collecting coverage data when using run()
93541035|2020-12-30T16:38:13+01:00|3721|3733|3770|Fix postgrest-coverage throwing error when installed with nix-env
519550e9|2020-12-30T17:52:15+01:00|3675|3759|3752|Add codecov to README.md
cc54135b|2020-12-31T14:37:52+01:00|3768|3796|3817|Rename master branch to main; replace whiteList with isAllowed
af278846|2021-01-02T13:50:19+01:00|3674|3720|3823|Add basic error handling to --dump-schema
56a7bc50|2021-01-02T13:51:00+01:00|3763|3707|3726|nix: Fix build errors with stale .tix files present
94b516d0|2021-01-02T13:51:00+01:00|3701|3739|3752|Add flag to postgrest.cabal to disable collecting coverage data
eed6015e|2021-01-03T17:54:28+01:00|3596|3738|3715|refactor: Move addHasVariadic to SQL
6efa3041|2021-01-03T17:54:28+01:00|3751|3758|3755|refactor: Move parseArg to SQL
1edbef05|2021-01-03T17:54:28+01:00|3612|3744|3862|Add STABLE RPCs in test fixtures to increase coverage
45ebaf0a|2021-01-03T17:54:28+01:00|3767|3756|3745|cov: Add comment on SCHEMA in test fixtures
b8e6450a|2021-01-03T17:54:28+01:00|3795|3812|3778|refactor: simplify addXRels
d173b7d8|2021-01-03T17:54:28+01:00|3701|3728|3837|cov: remove unused fields from type Column
52230821|2021-01-03T17:54:28+01:00|3704|3703|3739|refactor: Combine relConstraint and relJunction in relLink
ab6f90aa|2021-01-04T23:14:44+01:00|3688|3737|3842|nix: Add postgrest-with-* tools to run with temporary databases of different versions
97d84563|2021-01-14T16:40:29+01:00|3729|3726|3846|cov: Add io tests for basic cli commands and invalid config options
6dd12646|2021-01-14T16:40:29+01:00|3712|3747|3728|cov: Remove unused code
ba86b479|2021-01-14T16:40:29+01:00|3699|3762|3781|cov: refactor ApiRequest profile header detection
15039553|2021-01-19T13:49:40-05:00|3687|3783|3738|Correct openapi preparing statements by default
67461504|2021-01-19T13:49:40-05:00|3697|3770|3755|Correct --dump-schema swallowing error
9c005fc6|2021-01-19T13:49:40-05:00|2253|2283|2293|feat: get configuration parameters from the db
125ea8f6|2021-01-19T13:49:40-05:00|2273|2275|2288|feat: allow reloading config with NOTIFY
4344cc92|2021-01-22T15:56:08-05:00|2279|2288|2319|refactor: use optString and move overrideFrom
17af56ad|2021-01-22T15:56:08-05:00|2291|2333|2301|refactor: config validation inside readAppConfig
6557f1f9|2021-01-22T15:56:08-05:00|2270|2299|2334|correct NOTIFY config reload dying on error
c93e8f9e|2021-01-25T18:38:46-05:00|2282|2287|2308|Correct hardcoded postgrest_test_authenticator (#1743)
0ddd676e|2021-02-07T22:15:44+01:00|2279|2292|2321|nix: Add a script to generate a graph of Haskell imports (#1751)
e6973f96|2021-02-23T22:41:48+01:00|2211|2252|2263|refactor: App.hs and related changes (#1725)
b50b6749|2021-02-25T20:03:38-05:00|2183|2262|2256|Update CYBERTEC logo and BACKERS
a7dab6d9|2021-02-27T13:26:02-05:00|2170|2238|2248|cirrusci: update to FreeBSD 12.2
c3ccaf1a|2021-02-28T13:02:37-05:00|2170|2247|2252|Allow lens-5.0
e4516ab6|2021-03-05T19:03:15-05:00|2221|2265|2269|"Correct db settings to use ""_"" instead of ""-"""
6750a5c4|2021-03-05T19:03:15-05:00|2199|2243|2242|Restrict db settings to current db and global
498e7721|2021-03-05T19:03:15-05:00|2195|2235|2242|Reread in-db config when recoverying connection
d3a8b5f6|2021-03-05T19:03:15-05:00|2202|2243|2249|Change db-load-guc-config to db-config
5baec481|2021-03-14T12:45:02-05:00|2207|2239|2231|circleci: update base image
3a466aea|2021-03-14T12:45:02-05:00|2202|2291|2240|tests: don't match JSON bodies literally
71f6061d|2021-03-25T20:42:43-05:00|2156|2236|2269|Add Supabase as a sponsor
9f074cec|2021-03-26T10:13:38-05:00|2210|2220|2243|cirrusci: update FreeBSD pkg
65e7f9e8|2021-04-06T11:46:12+02:00|2235|2234|2233|nix: Add a tool for checking Haskell imports and exports (#1768)
376beac2|2021-04-06T17:07:14+02:00|2191|2252|2273|update contributing guidelines
9c79a417|2021-04-06T22:22:13-05:00|2202|2216|2277|Add Contributor Covenant Code of Conduct v2
7f11c1a9|2021-04-08T01:45:13+02:00|2208|2279|2231|fix: Void functions return null instead of empty body (#1795)
67f555d2|2021-04-08T14:33:18+02:00|2202|2219|2248|nix: Fix incorrect use of bash variable (#1797)
a5372e47|2021-04-09T18:48:00+02:00|2194|2249|2229|nix: show better error messages for missing arguments in hsie scripts (#1798)
9118a4a7|2021-04-10T20:38:51+02:00|2206|2254|2243|fix random codecov/project drops in CI, resolves #1800 
8c44410c|2021-04-11T10:38:35+02:00|2234|2247|2258|remove misleadingly named emptyOnFalse and similar (#1803)
f99fd6cb|2021-04-11T18:28:01+02:00|2234|2257|2228|refactor: Split up Types.hs and logically organize modules (#1793)
579a626d|2021-04-12T19:03:16+02:00|2216|2246|2225|nix: move `withtmpdb` to separate file
fc791a63|2021-04-12T19:03:16+02:00|2212|2238|2269|nix: add `withTmpDir` to `checked-shell-script`
67ca814d|2021-04-12T19:03:16+02:00|2208|2243|2237|nix: properly throw away tix files for `postgrest-watch`, resolves #1807 
801e229c|2021-04-12T19:17:47+02:00|2159|2216|2211|refactor: Carve CLI and Version modules out of Config (#1806)
91d0c732|2021-04-12T20:31:32+02:00|2217|2239|2256|add comment to Setup.hs
d40d7040|2021-04-12T20:31:32+02:00|2196|2234|2251|remove unused static assets
c2c7bbe9|2021-04-12T20:47:22+02:00|2207|2218|2204|refactor: remove obsoleted Private module (#1811)
1db32e2b|2021-04-12T21:01:15+02:00|2182|2208|2212|nix: add `postgrest-loadtest`
e44537ae|2021-04-12T21:01:15+02:00|2196|2244|2263|nix: small refactors
c502756e|2021-04-13T17:26:19+02:00|2212|2202|2218|add todo
bbcf942f|2021-04-13T17:26:33+02:00|2228|2235|2234|add a few header tests
4172a72b|2021-04-13T21:15:40+02:00|2170|2243|2259|legacy support

</details>

And here's how that looks:

![loadtest](https://user-images.githubusercontent.com/9132420/114680418-764fa980-9d0d-11eb-8717-da1195cf65db.png)


The commits with the major changes are (first number is position on x-axis):
- 50: d1d0c677 perf: change Text queries to ByteString
- 89: 4bd5e6bd perf: enable prepared statements for GET
- 105: 7069bb3c refactor: Change SET LOCAL gucs to set_config
- 106: 11d62a80 Prepared statement for set_config
- 127: c7f0d423 Add postgrest-coverage to show and upload hpc reports to codecov
- 148: 9c005fc6 feat: get configuration parameters from the db

The first 2 are major improvements, but we knew about those before. The two `set_config` ones are funny - luckily we only had that regression for 1 commit and then introduced prepared statements, too! :)

The coverage-one is not critical. This drop in performance is not real (for production). That's just the effect of changing cabal's build flags to `-O0` and enabling HPC coverage all the time to avoid rebuilds. This only happens in development. As a side-note: I think it's actually beneficial, because it seems like before that change the numbers were much noisier. I assume that vegeta on 1 core was running at it's limits at that point. By making PostgREST slower in general, vegeta will still be able to saturate it.

But.. the last one is really surprising. I have no idea why the addition of configuration parameters from the database slows things down so much. This is something we need to investigate.

@steve-chavez, could you do me a favor and test the following commits vs. the commit immediately before with your k6-setup?

- c7f0d423 - to confirm that the drop related to coverage is really only a development thing. You'd need to make sure you run a production build, but I guess you do anyway.
- 9c005fc6 - to see how much that drop is in "real" numbers. The hypothetical req/s are exactly that... hypothetical.

---

Another observation: The blue numbers (1s) are a bit jumpy, I feel. But there is not too big of a difference between orange and green it seems. So it should be enough to run each test-case for 5s and not 25s.

steve-chavez commented:
:O! Nice plotting work Wolfgang.

> could you do me a favor and test the following commits vs. the commit immediately before with your k6-setup?

Sure! I'll create static binaries for those commits and report here. 

Meanwhile, I've ran [GETSingle](https://github.com/supabase/benchmarks/blob/master/postgrest/k6/GETSingle.js#L58-L59) on the latest nightly(https://github.com/PostgREST/postgrest/commit/d3a8b5f6e1f541c5c6cde9398a7ba6a95cdbdb88) — which already contains the coverage and in-db config changes — and v7.0.1.

Nightly(`2071.345877/s`):
```
    data_received..............: 20 MB  665 kB/s
    data_sent..................: 6.5 MB 217 kB/s
  ✓ dropped_iterations.........: 750    24.955574/s
  ✓ failed requests............: 0.00%  ✓ 0     ✗ 62251
    http_req_blocked...........: avg=8.52µs   min=1.79µs   med=3.3µs   max=55.83ms  p(90)=4.5µs   p(95)=5.94µs
    http_req_connecting........: avg=2.57µs   min=0s       med=0s      max=32.5ms   p(90)=0s      p(95)=0s
  ✓ http_req_duration..........: avg=7.5ms    min=795.97µs med=2.05ms  max=150.47ms p(90)=25.32ms p(95)=39ms
    http_req_receiving.........: avg=109.1µs  min=12.87µs  med=36.42µs max=41.28ms  p(90)=81.51µs p(95)=179.47µs
    http_req_sending...........: avg=327.89µs min=7.32µs   med=27.77µs max=47.78ms  p(90)=48.94µs p(95)=109.54µs
    http_req_tls_handshaking...: avg=0s       min=0s       med=0s      max=0s       p(90)=0s      p(95)=0s
    http_req_waiting...........: avg=7.06ms   min=753.68µs med=1.98ms  max=150.36ms p(90)=23.5ms  p(95)=37.03ms
    http_reqs..................: 62251  2071.345877/s
    iteration_duration.........: avg=7.64ms   min=889.53µs med=2.18ms  max=150.61ms p(90)=25.49ms p(95)=39.2ms
    iterations.................: 62251  2071.345877/s
    vus........................: 121    min=107 max=121
    vus_max....................: 121    min=107 max=121
```
v7.0.1(`1347.378153/s`):
```
    data_received..............: 12 MB  411 kB/s
    data_sent..................: 4.3 MB 141 kB/s
  ✗ dropped_iterations.........: 22213  734.732071/s
  ✓ failed requests............: 0.00%  ✓ 0     ✗ 40735
    http_req_blocked...........: avg=105.85µs min=1.47µs  med=3.79µs   max=166.33ms p(90)=6.57µs   p(95)=15.15µs
    http_req_connecting........: avg=50.66µs  min=0s      med=0s       max=165.71ms p(90)=0s       p(95)=0s
  ✓ http_req_duration..........: avg=207.25ms min=1.64ms  med=199.9ms  max=710.8ms  p(90)=351.09ms p(95)=412.99ms
    http_req_receiving.........: avg=891.97µs min=13.56µs med=35.72µs  max=174.84ms p(90)=324.3µs  p(95)=2.04ms
    http_req_sending...........: avg=12.83ms  min=7.95µs  med=29.88µs  max=272.68ms p(90)=49.77ms  p(95)=87.19ms
    http_req_tls_handshaking...: avg=0s       min=0s      med=0s       max=0s       p(90)=0s       p(95)=0s
    http_req_waiting...........: avg=193.52ms min=1.56ms  med=193.4ms  max=629.33ms p(90)=320.26ms p(95)=352.7ms
    http_reqs..................: 40735  1347.378153/s
    iteration_duration.........: avg=209.56ms min=1.8ms   med=200.23ms max=783.64ms p(90)=358.69ms p(95)=422.12ms
    iterations.................: 40735  1347.378153/s
    vus........................: 514    min=110 max=514
    vus_max....................: 514    min=110 max=514
```
Seems there's no regression.

steve-chavez commented:
> c7f0d42 - to confirm that the drop related to coverage is really only a development thing

2061.35802/s - https://github.com/PostgREST/postgrest/commit/201568831267441297c0906c3bd1c0a7a7378b06 - Fix spec test concurrency issue related to tx=commit on simple_pk
```
    data_received..............: 20 MB  655 kB/s
    data_sent..................: 6.5 MB 216 kB/s
  ✓ dropped_iterations.........: 925    30.716974/s
  ✓ failed requests............: 0.00%  ✓ 0     ✗ 62075
    http_req_blocked...........: avg=8.64µs   min=1.61µs   med=3.54µs  max=20.46ms  p(90)=4.83µs  p(95)=6.23µs
    http_req_connecting........: avg=2.79µs   min=0s       med=0s      max=20.34ms  p(90)=0s      p(95)=0s
  ✓ http_req_duration..........: avg=9.59ms   min=809.9µs  med=2.19ms  max=201.53ms p(90)=33.22ms p(95)=47.23ms
    http_req_receiving.........: avg=115.67µs min=12.3µs   med=37.03µs max=52.77ms  p(90)=85.91µs p(95)=191.4µs
    http_req_sending...........: avg=365.3µs  min=7.59µs   med=28.25µs max=47.23ms  p(90)=50.24µs p(95)=128.92µs
    http_req_tls_handshaking...: avg=0s       min=0s       med=0s      max=0s       p(90)=0s      p(95)=0s
    http_req_waiting...........: avg=9.11ms   min=765.52µs med=2.1ms   max=201.46ms p(90)=31.58ms p(95)=45.97ms
    http_reqs..................: 62075  2061.35802/s
    iteration_duration.........: avg=9.74ms   min=937.83µs med=2.33ms  max=202.04ms p(90)=33.4ms  p(95)=47.39ms
    iterations.................: 62075  2061.35802/s
    vus........................: 133    min=108 max=133
    vus_max....................: 133    min=108 max=133
```

2039.801516/s - https://github.com/PostgREST/postgrest/commit/c7f0d423238a70c23a82ed136fc18e1843c007ca -  Add postgrest-coverage to show and upload hpc reports to codecov

```
    data_received..............: 20 MB  649 kB/s
    data_sent..................: 6.4 MB 213 kB/s
  ✓ dropped_iterations.........: 1418   46.95822/s
  ✓ failed requests............: 0.00%  ✓ 0     ✗ 61596
    http_req_blocked...........: avg=8.6µs    min=1.58µs   med=3.23µs  max=43.54ms  p(90)=4.43µs  p(95)=5.74µs
    http_req_connecting........: avg=3.8µs    min=0s       med=0s      max=43.35ms  p(90)=0s      p(95)=0s
  ✓ http_req_duration..........: avg=10.8ms   min=820.23µs med=2.25ms  max=201.15ms p(90)=39.68ms p(95)=55.87ms
    http_req_receiving.........: avg=136.35µs min=11.61µs  med=35.39µs max=101.31ms p(90)=79.13µs p(95)=184.77µs
    http_req_sending...........: avg=580.15µs min=7.28µs   med=27.49µs max=60.75ms  p(90)=52.18µs p(95)=167.99µs
    http_req_tls_handshaking...: avg=0s       min=0s       med=0s      max=0s       p(90)=0s      p(95)=0s
    http_req_waiting...........: avg=10.09ms  min=771.72µs med=2.16ms  max=166.73ms p(90)=36.34ms p(95)=53.09ms
    http_reqs..................: 61596  2039.801516/s
    iteration_duration.........: avg=10.95ms  min=920.37µs med=2.38ms  max=201.32ms p(90)=39.83ms p(95)=56.05ms
    iterations.................: 61596  2039.801516/s
    vus........................: 139    min=108 max=139
    vus_max....................: 139    min=108 max=139
```

wolfgangwalther commented:
I have done a few further tests and checks:
- When I disable the development / hpc flags, the coverage-drop is gone as expected. But the config-drop is still there - down to ~3000 hypo/s.
- The same shows for all metrics I collect (mean, median, min). Actual req/s in my test-setup go down from ~2200 to ~1400 with the config commit. This does not seem to be an issue with the metrics I am using.

wolfgangwalther commented:
@steve-chavez In your setup: Which parts of the setup (postgrest, database, workers) are on the same machine? All three? Or are any of them going through a real network?

steve-chavez commented:
postgrest + database are on the same machine(a t3a.nano, low-end instance) the k6 runner is in another machine(t3a.medium). The two machines are under the same VPC(AWS's VLAN). So yep, the requests go through network.

wolfgangwalther commented:
Figured it out:
- When I disable `db-config`, the regression is gone.
- Because we're running the loadtest with the test fixtures schema, we also got a few settings of `postgrest_test_authenticator`. Namely:
  - `ALTER ROLE postgrest_test_authenticator SET pgrst.db_prepared_statements = 'false';`
  - `ALTER ROLE postgrest_test_authenticator SET pgrst.db_pre_request = 'test.custom_headers';`

Just by removing those two lines, I am up at 5200 hypo/s again (including the disabled dev flags).

Soo.... we should either disable `db-config` for those loadtests or use a completely different schema / user etc. right away.

steve-chavez commented:
> ALTER ROLE postgrest_test_authenticator SET pgrst.db_prepared_statements = 'false';
> ALTER ROLE postgrest_test_authenticator SET pgrst.db_pre_request = 'test.custom_headers';

Aha! Disabling the prepared statements should have the most impact. Btw, regarding `pre-request`, I also noted it impacts perf. My results with a simple prereq that did nothing but return false were `1844.326975/s`(around `11%` perf loss) on the same GETSingle test.

wolfgangwalther commented:
> Aha! Disabling the prepared statements should have the most impact. Btw, regarding `pre-request`, I also noted it impacts perf. My results with a simple prereq that did nothing but return false were `1844.326975/s`(around `11%` perf loss) on the same GETSingle test.

Yes, enabling prepared statements gave me the biggest boost. That 11% drop is probably that extra round-trip on every request then.

> **Edit**: Added the 4 tests on [#1812 (comment)](https://github.com/PostgREST/postgrest/pull/1812#issuecomment-819660198).

Thanks. I'm glad that we're still running fast :)

wolfgangwalther commented:
@monacoremo could you have a look at my latest commit 2b71f53? I hope that will fix the "missing GHC" error I got in CI. What do you think about the way I did that - any nix-anti-pattern or something that I hit? ;)

monacoremo commented on nix/overlays/checked-shell-script/checked-shell-script.nix:
```diff
@@ -50,6 +56,10 @@ let
 
+        + lib.optionalString (useEnv != false) ''
+          export PATH="${useEnv.env}/bin:$PATH"
+        ''
```
This should be

```
env="$(cat ${useEnv.env})"
export PATH="$env/bin:$PATH"
```

The `$(cat ...)` part is significant

wolfgangwalther replied:
I just realized that, too. :D

(latest force-push changed just that)

monacoremo replied:
The name `useEnv` seems a bit off... maybe `withEnv` or `withEnvOf` with a default of `null`?

monacoremo replied:
Cleanest option probably is `withEnv` with a default of `null` and pass in `postgrest.env` on the call sites

monacoremo replied:
@wolfgangwalther unresolving to make sure you saw the naming :)

wolfgangwalther replied:
I went with that, much better. Will wait to push until the current CI run finishes, to hopefully get the first two CI runs in already.

wolfgangwalther commented:
@monacoremo another question for, maybe you already have an idea. I am now at a stage where I want to parse not only arguments, but options to those shell scripts (loadtest and changelog stuff, too). I thought I'd get my feet wet with the "do something with `docs` TODO" we have. I have looked at `getopt` and  `getopts` so far. Not terribly convinced, yet. Do you know anything else that could help here? Or have any experience with either of the two?

wolfgangwalther commented:
A few runs of `postgrest-loadtest` in CI. It runs a comparison against the main branch, but since we don't have any source-code changes in this branch, we should expect the same result.

First run:

```bash
Running loadtest on "main"...

Requests      [total, rate, throughput]         5698, 1139.48, 1139.25
Duration      [total, attack, wait]             5.002s, 5.001s, 1.011ms
Latencies     [min, mean, 50, 90, 95, 99, max]  640.089µs, 873.787µs, 802.365µs, 1.131ms, 1.236ms, 1.513ms, 3.974ms
Bytes In      [total, mean]                     239316, 42.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:5698  
Error Set:

Running loadtest on HEAD...

Requests      [total, rate, throughput]         5549, 1109.72, 1109.38
Duration      [total, attack, wait]             5.002s, 5s, 1.562ms
Latencies     [min, mean, 50, 90, 95, 99, max]  592.531µs, 897.615µs, 827.983µs, 1.172ms, 1.265ms, 1.504ms, 4.553ms
Bytes In      [total, mean]                     233058, 42.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:5549  
Error Set:
```

wolfgangwalther commented:
> @monacoremo another question for, maybe you already have an idea. I am now at a stage where I want to parse not only arguments, but options to those shell scripts (loadtest and changelog stuff, too). I thought I'd get my feet wet with the "do something with `docs` TODO" we have. I have looked at `getopt` and `getopts` so far. Not terribly convinced, yet. Do you know anything else that could help here? Or have any experience with either of the two?

I have just found https://argbash.io. Looks like we could integrate that into our `checked-shell-script`. This way we could provide an additional argument `args` or something that holds the argbash template. We could then add the `--help` option in checked-shell-script to that and let argbash to the rest. WDYT?

monacoremo commented:
> I have just found https://argbash.io.

Looks super promising! That would bring all our scripts to the next level!

I'm only familiar with `optparse-applicative` in Haskell and `click` in Python. But this could be awesome for our bash scripts!

wolfgangwalther commented:
To be able to run the loadtest consistently against different commits, we need to not only run the same requests, but also run them on the same schema. Assuming some PR makes a change and adds a loadtest to show the result - the same test, including fixtures, needs to be run against the main branch. To avoid a complicated procedure of needing to commit tests/fixtures first, etc. I was looking for a way to tell `withTmpDb` to load a different set of fixtures - not the spec-test fixtures.

To do this, I wanted to add CLI option to `with_tmp_db` and then felt the best way to do that, was to make it a checked-shell-script. Yes, this removes the ability to run it without nix, but I felt it was worth it, because:
- There is still `test/create_test_db`
- One could install `postgrest-with-xxx` via nix-env and then still run stack tests against it (or do it inside nix-shell)
- It makes the script significantly shorter, because of features already included in checkedShellScript
- It allows us to add the `--fixtures` CLI option easily

@monacoremo WDYT?

---

Now, I added a very small schema for the loadtest and also more requests to test. My first goal was to match all request handlers in `App.hs`. Did I get all of them? What main features, that would touch more (or possibly performance critical) code, should we add?

monacoremo commented:
> To do this, I wanted to add CLI option to `with_tmp_db` and then felt the best way to do that, was to make it a checked-shell-script. Yes, this removes the ability to run it without nix, but I felt it was worth it

Yes, I see how that would be nice - especially with the argbash integration now. `test/create_test_db` is a bit of a grown mess, would be good if we could phase it out actually...

I was thinking whether there could be a conpromize - e.g. generating and checking in a with_tmp_db based on the nix version. Should be doable with a bit of sed (changing the shebang, removing nix-store refences, embedding the arg parsing, adding a comment that the file is generated etc) and we could check the consistency in CI as we do with `style`. Seems like a lot of extra work though. We should then also move the stack tests to with_tmp_db and kill create_test_db completely.

Or we just maintain two versions of the script, it's been relatively stable...

WDYT?

wolfgangwalther commented:
> We should keep some version of the standalone script in any case I think, otherwise we seriously gimp any non-nix developers.

I don't think there is a reason to do that. I can see the reasoning for allowing to build with stack. But not without nix. All our dev environment is based on nix + we have the docker-compose setup for cases where that doesn't work natively. To be honest: The biggest hurdle for me going up to speed initially was the fact that I was **not** using nix. Once I did, everything was so much smoother.

Let's turn this around and create a `postgrest-build-stack` command **in** nix. And then we keep **everything** in nix.

wolfgangwalther commented:
I rebased this after the move to GitHub Actions. Right now, `cachix/install-nix-action@v13` seems to have some problems, see e.g.:

https://github.com/PostgREST/postgrest/runs/4228318238?check_suite_focus=true

Makes all the nix based jobs fail randomly.

---

The load test job is supposed to upload a markdown table to the checks page of the PR. This is not working right now, because PRs from a forked repo only have read access through `secrets.GITHUB_TOKEN`. I'll need to pass the markdown file as an artifact to a trusted workflow and then upload it from there, as mentioned here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/.

wolfgangwalther commented:
So the loadtest report seems to be uploaded successfully, according to the workflow:
https://github.com/PostgREST/postgrest/runs/4317690798?check_suite_focus=true

Unfortunately I have no idea where it ended up. Can't find it anywhere, yet. Will have to investigate later..

monacoremo commented:
> Unfortunately I have no idea where it ended up. Can't find it anywhere, yet. Will have to investigate later..

Something turned up here!

![Screenshot_20211125-230511_Chrome](https://user-images.githubusercontent.com/59358383/143503549-6ccd5ebb-dd71-4b78-a34c-ea2785da1a7c.jpg)

https://github.com/PostgREST/postgrest/runs/4317691924?check_suite_focus=true

Need to check whether that went to the correct check run, but seems to work great otherwise. Very stable results also, awesome work!

wolfgangwalther commented:
> Something turned up here!
> [...]
> Need to check whether that went to the correct check run, [...]

Ah nice find. The workflow that runs on `workflow_run` is running on the latest commit of the default branch - that's why the report went to that commits check run. I just pushed a fix for that, it should now target the correct commit.

Let's see what happens in a bit.

wolfgangwalther commented:
Ok, so that worked fine now, it shows up here. Right now the results are only added with a neutral conclusion. There is no comparison against a threshold, yet, and the pipeline would not fail because of a drop in performance.

We can improve the formatting of the results later on, too - and also write a more sophisticated python script to analyze the numbers in detail, e.g. separated by request. Right now we just a have one set of numbers for all kinds of requests mixed together.

My idea would be to merge this now, and then let this run a bit on all pull requests and pushes to main. We can then get a feel for the numbers and see what kind of threshold we would like to implement.

WDYT?

monacoremo commented on nix/tools/withTools.nix:
```diff
@@ -113,21 +118,159 @@ let
 {
   name = "postgrest-with";
-  tools = [ withAll ] ++ withVersions;
+  tools = [ withAll withGit withPgrst ] ++ withVersions;
```
With the extended scope of withTools, we should precise `withAll` a bit. E.g. `withPgAll` or `withTmpDbAllVersions`? Would also be good to be consistent with naming the other attributes, why not just `inherit withGit withPgrst`?

wolfgangwalther replied:
Yes, makes sense. I improved the naming!

monacoremo reviewed (approved):
Amazing work! Great integration with our CI, will be very useful to have
